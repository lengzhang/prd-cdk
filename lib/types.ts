import * as cdk from "aws-cdk-lib";

export interface Environment extends cdk.Environment {
  stage: string;
}
