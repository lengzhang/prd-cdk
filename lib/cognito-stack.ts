import * as cdk from "aws-cdk-lib";
import * as cognito from "aws-cdk-lib/aws-cognito";
import * as cognitoIdentityPool from "@aws-cdk/aws-cognito-identitypool-alpha";
import * as iam from "aws-cdk-lib/aws-iam";
import { Construct } from "constructs";

interface CognitoStackProps extends cdk.StackProps {
  userPoolName: string;
  userPoolClientName: string;
  userPoolDomainName: string;
  userPoolDomainPrefix: string;
  identityPoolName: string;
  reportsBucketName: string;
}

class CognitoStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CognitoStackProps) {
    super(scope, id, props);

    const { userPool, userPoolClient } = this.createUserPool(
      props.userPoolName,
      props.userPoolClientName,
      props.userPoolDomainName,
      props.userPoolDomainPrefix
    );

    this.createIdentityPool(
      props.identityPoolName,
      userPool,
      userPoolClient,
      props.reportsBucketName
    );
  }

  createUserPool(
    userPoolName: string,
    userPoolClientName: string,
    userPoolDomainName: string,
    userPoolDomainPrefix: string
  ) {
    const userVerification: cognito.UserVerificationConfig = {
      emailSubject: "[PRD] You need to verify your email",
      emailBody:
        "Thanks for signing up. Please verify your account by the verification code {####}.",
      emailStyle: cognito.VerificationEmailStyle.CODE,
    };
    const standardAttributes: cognito.StandardAttributes = {
      email: { mutable: false, required: true },
      preferredUsername: { mutable: true, required: true },
      lastUpdateTime: { mutable: false, required: true },
    };

    const customAttributes: {
      [key: string]: cdk.aws_cognito.ICustomAttribute;
    } = {
      // timestamp for date time
      created_at: new cognito.NumberAttribute({ mutable: false }),
    };

    const passwordPolicy: cognito.PasswordPolicy = {
      minLength: 8,
      requireLowercase: true,
      requireUppercase: true,
      requireDigits: true,
      requireSymbols: false,
    };

    const userPool = new cognito.UserPool(this, userPoolName, {
      userPoolName,
      signInAliases: { email: true },
      selfSignUpEnabled: true,
      autoVerify: { email: true },
      userVerification,
      standardAttributes,
      customAttributes,
      passwordPolicy,
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    userPool.addDomain(userPoolDomainName, {
      cognitoDomain: {
        domainPrefix: userPoolDomainPrefix,
      },
    });

    const userPoolClient = userPool.addClient(userPoolClientName, {
      userPoolClientName: userPoolClientName,
      authFlows: {
        userPassword: true,
      },
    });

    return { userPool, userPoolClient };
  }

  createIdentityPool(
    identityPoolName: string,
    userPool: cognito.UserPool,
    userPoolClient: cognito.UserPoolClient,
    reportBucketName: string
  ) {
    const userPoolAuthenticationProvider =
      new cognitoIdentityPool.UserPoolAuthenticationProvider({
        userPool,
        userPoolClient,
      });
    const identityPool = new cognitoIdentityPool.IdentityPool(
      this,
      identityPoolName,
      { identityPoolName }
    );

    identityPool.addUserPoolAuthentication(userPoolAuthenticationProvider);
    identityPool.applyRemovalPolicy(cdk.RemovalPolicy.DESTROY);

    const reportBucketAccessPolicy = new iam.PolicyStatement({
      effect: iam.Effect.ALLOW,
      actions: ["s3:GetObject", "s3:ListBucket", "s3:PutObject"],
      resources: [
        `arn:aws:s3:::${reportBucketName}`,
        `arn:aws:s3:::${reportBucketName}/*`,
      ],
    });
    identityPool.authenticatedRole.addToPrincipalPolicy(
      reportBucketAccessPolicy
    );
  }
}

export default CognitoStack;
