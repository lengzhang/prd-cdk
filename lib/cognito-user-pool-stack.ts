import * as cdk from "aws-cdk-lib";
import * as cognito from "aws-cdk-lib/aws-cognito";
import { Construct } from "constructs";

interface CognitoUserPoolStackProps extends cdk.StackProps {
  userPoolName: string;
  appClientName: string;
  cognitoDomainName: string;
  cognitoDomainPrefix: string;
}

class CognitoUserPoolStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: CognitoUserPoolStackProps) {
    super(scope, id, props);

    this.createUserPool(
      props.userPoolName,
      props.appClientName,
      props.cognitoDomainName,
      props.cognitoDomainPrefix
    );
  }

  createUserPool(
    userPoolName: string,
    appClientName: string,
    cognitoDomainName: string,
    cognitoDomainPrefix: string
  ) {
    const userVerification: cognito.UserVerificationConfig = {
      emailSubject: "[PRD] You need to verify your email",
      emailBody:
        "Thanks for signing up. Please verify your account by the verification code {####}.",
      emailStyle: cognito.VerificationEmailStyle.CODE,
    };
    const standardAttributes: cognito.StandardAttributes = {
      email: { mutable: false, required: true },
      familyName: { mutable: true, required: true },
      givenName: { mutable: true, required: true },
      preferredUsername: { mutable: true, required: true },
      lastUpdateTime: { mutable: false, required: true },
    };

    const customAttributes: {
      [key: string]: cdk.aws_cognito.ICustomAttribute;
    } = {
      // timestamp for date time
      created_at: new cognito.NumberAttribute({ mutable: false }),
    };

    const passwordPolicy: cognito.PasswordPolicy = {
      minLength: 8,
      requireLowercase: true,
      requireUppercase: true,
      requireDigits: true,
      requireSymbols: false,
    };

    const userPool = new cognito.UserPool(this, userPoolName, {
      userPoolName,
      signInAliases: { email: true },
      selfSignUpEnabled: true,
      autoVerify: { email: true },
      userVerification,
      standardAttributes,
      customAttributes,
      passwordPolicy,
      accountRecovery: cognito.AccountRecovery.EMAIL_ONLY,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
    });

    userPool.addDomain(cognitoDomainName, {
      cognitoDomain: {
        domainPrefix: cognitoDomainPrefix,
      },
    });

    userPool.addClient(appClientName, {
      userPoolClientName: appClientName,
      authFlows: {
        userPassword: true,
      },
    });
  }
}

export default CognitoUserPoolStack;
