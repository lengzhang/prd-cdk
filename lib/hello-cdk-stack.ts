import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";

interface HelloCdkStackProps extends cdk.StackProps {
  bucketName: string;
}

class HelloCdkStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: HelloCdkStackProps) {
    super(scope, id, props);

    new cdk.aws_s3.Bucket(this, props.bucketName, {
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });
  }
}

export default HelloCdkStack;
