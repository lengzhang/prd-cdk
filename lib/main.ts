#!/usr/bin/env node
import "source-map-support/register";
import * as dotenv from "dotenv";

import * as cdk from "aws-cdk-lib";

import PrdResourcesStack from "./prd-resources-stack";
import CognitoUserPoolStack from "./cognito-stack";
import { Environment } from "./types";

dotenv.config();

const appName = "prd";
const env: Environment = {
  account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
  stage: process.env.STAGE || "",
};

function main() {
  const app = new cdk.App();
  const namePrefix = `${env.stage}-${appName}-${env.account}-${env.region}`;

  const prdResourcesStackName = `${namePrefix}-resources-stack`;
  const prdResourcesBucketName = `${namePrefix}-reports-bucket`;
  new PrdResourcesStack(app, prdResourcesStackName, {
    env,
    reportsBucketName: prdResourcesBucketName,
  });

  const cognitoStackName = `${namePrefix}-cognito-stack`;
  const userPoolName = `${namePrefix}-cognito-user-pool`;
  const userPoolClientName = `${namePrefix}-cognito-user-pool-client`;
  const userPoolDomainName = `${namePrefix}-cognito-user-pool-domain`;
  const userPoolDomainPrefix = `${env.stage}-${appName}-${env.account}`;
  const identityPoolName = `${namePrefix}-cognito-identity-pool`;
  new CognitoUserPoolStack(app, cognitoStackName, {
    env,
    userPoolName,
    userPoolClientName,
    userPoolDomainName,
    userPoolDomainPrefix,
    identityPoolName,
    reportsBucketName: prdResourcesBucketName,
  });
}

main();
