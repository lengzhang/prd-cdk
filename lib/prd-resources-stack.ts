import * as cdk from "aws-cdk-lib";
import * as s3 from "aws-cdk-lib/aws-s3";
import { Construct } from "constructs";

import { Environment } from "./types";

interface PrdResourcesStackProps extends cdk.StackProps {
  env: Environment;
  reportsBucketName: string;
}

class PrdResourcesStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: PrdResourcesStackProps) {
    super(scope, id, props);
    this.createPrdReportsBucket(props.reportsBucketName, props.env);
  }

  createPrdReportsBucket(reportsBucketName: string, env: Environment) {
    const cors: s3.CorsRule[] = [];
    const corsRule: s3.CorsRule = {
      allowedMethods: [s3.HttpMethods.GET, s3.HttpMethods.PUT],
      allowedOrigins: [],
      allowedHeaders: ["*"],
    };
    if (env.stage === "prod" && process.env.ORIGIN) {
      corsRule.allowedOrigins.push(process.env.ORIGIN);
      cors.push(corsRule);
    } else if (env.stage === "dev") {
      corsRule.allowedOrigins.push("*");
      cors.push(corsRule);
    }

    new s3.Bucket(this, reportsBucketName, {
      bucketName: reportsBucketName,
      versioned: true,
      removalPolicy: cdk.RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
      cors,
    });
  }
}

export default PrdResourcesStack;
