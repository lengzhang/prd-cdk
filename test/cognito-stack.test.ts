import * as cdk from "aws-cdk-lib";
import { Template, Match } from "aws-cdk-lib/assertions";

import CognitoStack from "../lib/cognito-stack";
import { Environment } from "../lib/types";

const env: Environment = {
  account: process.env.CDK_DEPLOY_ACCOUNT || process.env.CDK_DEFAULT_ACCOUNT,
  region: process.env.CDK_DEPLOY_REGION || process.env.CDK_DEFAULT_REGION,
  stage: "test",
};

describe("CognitoStack", () => {
  const testStackName = "test-cognito-user-pool-stack";
  const testUserPoolName = "test-user-pool-name";
  const testUserPoolClientName = "test-user-pool-client-name";
  const testUserPoolDomainName = "test-user-pool-domain-name";
  const testUserPoolDomainPrefix = "test-user-pool-domain-prefix";
  const testIdentityPoolName = "test-identity-pool-name";
  const testReportsBucketName = "test-report-bucket-name";
  const getCognitoStack = () => {
    const app = new cdk.App();
    const stack = new CognitoStack(app, "test-cognito-user-pool-stack", {
      env,
      userPoolName: testUserPoolName,
      userPoolClientName: testUserPoolClientName,
      userPoolDomainName: testUserPoolDomainName,
      userPoolDomainPrefix: testUserPoolDomainPrefix,
      identityPoolName: testIdentityPoolName,
      reportsBucketName: testReportsBucketName,
    });
    return stack;
  };

  const stack = getCognitoStack();
  const template = Template.fromStack(stack);

  test("matches the snapshot", () => {
    expect(template.toJSON()).toMatchSnapshot();
    expect(stack.stackName).toBe(testStackName);
    template.hasResourceProperties("AWS::Cognito::UserPool", {
      UserPoolName: Match.exact(testUserPoolName),
    });
  });

  describe("UserPool", () => {
    test("check user verification config", () => {
      template.hasResourceProperties("AWS::Cognito::UserPool", {
        VerificationMessageTemplate: Match.objectEquals({
          DefaultEmailOption: "CONFIRM_WITH_CODE",
          EmailMessage:
            "Thanks for signing up. Please verify your account by the verification code {####}.",
          EmailSubject: "[PRD] You need to verify your email",
          SmsMessage: "The verification code to your new account is {####}",
        }),
      });
    });

    test("check standard and custom attributes", () => {
      const stack = getCognitoStack();

      const template = Template.fromStack(stack);

      template.hasResourceProperties("AWS::Cognito::UserPool", {
        Schema: Match.arrayEquals([
          // standard attributes
          { Mutable: false, Name: "email", Required: true },
          { Mutable: true, Name: "preferred_username", Required: true },
          { Mutable: false, Name: "updated_at", Required: true },
          // custom attributes
          { AttributeDataType: "Number", Mutable: false, Name: "created_at" },
        ]),
      });
    });

    test("check password policy", () => {
      const stack = getCognitoStack();

      const template = Template.fromStack(stack);

      template.hasResourceProperties("AWS::Cognito::UserPool", {
        Policies: {
          PasswordPolicy: Match.objectEquals({
            MinimumLength: 8,
            RequireLowercase: true,
            RequireNumbers: true,
            RequireSymbols: false,
            RequireUppercase: true,
          }),
        },
      });
    });

    test("check user pool domain", () => {
      template.hasResourceProperties("AWS::Cognito::UserPoolDomain", {
        Domain: Match.exact(testUserPoolDomainPrefix),
        UserPoolId: Match.objectEquals({
          Ref: Match.stringLikeRegexp(
            `^${testUserPoolName.replace(/\-/g, "")}`
          ),
        }),
      });
    });

    test("check user pool client", () => {
      template.hasResourceProperties("AWS::Cognito::UserPoolClient", {
        ClientName: Match.exact(testUserPoolClientName),
        UserPoolId: Match.objectEquals({
          Ref: Match.stringLikeRegexp(
            `^${testUserPoolName.replace(/\-/g, "")}`
          ),
        }),
        ExplicitAuthFlows: Match.arrayWith([
          Match.stringLikeRegexp("ALLOW_USER_PASSWORD_AUTH"),
        ]),
      });
    });
  });
});
