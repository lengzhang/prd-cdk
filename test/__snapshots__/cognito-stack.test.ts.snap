// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CognitoStack matches the snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "testidentitypoolname2BA432B1": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "testuserpoolname6617CE2F",
        "testuserpoolnametestuserpoolclientname68410DE1",
        "testuserpoolnametestuserpooldomainnameBB32991B",
      ],
      "Properties": {
        "AllowUnauthenticatedIdentities": false,
        "CognitoIdentityProviders": [
          {
            "ClientId": {
              "Ref": "testuserpoolnametestuserpoolclientname68410DE1",
            },
            "ProviderName": {
              "Fn::Join": [
                "",
                [
                  "cognito-idp.",
                  {
                    "Ref": "AWS::Region",
                  },
                  ".",
                  {
                    "Ref": "AWS::URLSuffix",
                  },
                  "/",
                  {
                    "Ref": "testuserpoolname6617CE2F",
                  },
                ],
              ],
            },
            "ServerSideTokenCheck": true,
          },
        ],
        "IdentityPoolName": "test-identity-pool-name",
      },
      "Type": "AWS::Cognito::IdentityPool",
      "UpdateReplacePolicy": "Delete",
    },
    "testidentitypoolnameAuthenticatedRole8340A8E2": {
      "DependsOn": [
        "testuserpoolname6617CE2F",
        "testuserpoolnametestuserpoolclientname68410DE1",
        "testuserpoolnametestuserpooldomainnameBB32991B",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "ForAnyValue:StringLike": {
                  "cognito-identity.amazonaws.com:amr": "authenticated",
                },
                "StringEquals": {
                  "cognito-identity.amazonaws.com:aud": {
                    "Ref": "testidentitypoolname2BA432B1",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Default Authenticated Role for Identity Pool ",
              {
                "Fn::GetAtt": [
                  "testidentitypoolname2BA432B1",
                  "Name",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testidentitypoolnameAuthenticatedRoleDefaultPolicy89557359": {
      "DependsOn": [
        "testuserpoolname6617CE2F",
        "testuserpoolnametestuserpoolclientname68410DE1",
        "testuserpoolnametestuserpooldomainnameBB32991B",
      ],
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:GetObject",
                "s3:ListBucket",
                "s3:PutObject",
              ],
              "Effect": "Allow",
              "Resource": [
                "arn:aws:s3:::test-report-bucket-name",
                "arn:aws:s3:::test-report-bucket-name/*",
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "testidentitypoolnameAuthenticatedRoleDefaultPolicy89557359",
        "Roles": [
          {
            "Ref": "testidentitypoolnameAuthenticatedRole8340A8E2",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "testidentitypoolnameDefaultRoleAttachmentD4F7ACA5": {
      "DependsOn": [
        "testuserpoolname6617CE2F",
        "testuserpoolnametestuserpoolclientname68410DE1",
        "testuserpoolnametestuserpooldomainnameBB32991B",
      ],
      "Properties": {
        "IdentityPoolId": {
          "Ref": "testidentitypoolname2BA432B1",
        },
        "Roles": {
          "authenticated": {
            "Fn::GetAtt": [
              "testidentitypoolnameAuthenticatedRole8340A8E2",
              "Arn",
            ],
          },
          "unauthenticated": {
            "Fn::GetAtt": [
              "testidentitypoolnameUnauthenticatedRole1BCD713F",
              "Arn",
            ],
          },
        },
      },
      "Type": "AWS::Cognito::IdentityPoolRoleAttachment",
    },
    "testidentitypoolnameUnauthenticatedRole1BCD713F": {
      "DependsOn": [
        "testuserpoolname6617CE2F",
        "testuserpoolnametestuserpoolclientname68410DE1",
        "testuserpoolnametestuserpooldomainnameBB32991B",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRoleWithWebIdentity",
              "Condition": {
                "ForAnyValue:StringLike": {
                  "cognito-identity.amazonaws.com:amr": "unauthenticated",
                },
                "StringEquals": {
                  "cognito-identity.amazonaws.com:aud": {
                    "Ref": "testidentitypoolname2BA432B1",
                  },
                },
              },
              "Effect": "Allow",
              "Principal": {
                "Federated": "cognito-identity.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Default Unauthenticated Role for Identity Pool ",
              {
                "Fn::GetAtt": [
                  "testidentitypoolname2BA432B1",
                  "Name",
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "testuserpoolname6617CE2F": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "AccountRecoverySetting": {
          "RecoveryMechanisms": [
            {
              "Name": "verified_email",
              "Priority": 1,
            },
          ],
        },
        "AdminCreateUserConfig": {
          "AllowAdminCreateUserOnly": false,
        },
        "AutoVerifiedAttributes": [
          "email",
        ],
        "EmailVerificationMessage": "Thanks for signing up. Please verify your account by the verification code {####}.",
        "EmailVerificationSubject": "[PRD] You need to verify your email",
        "Policies": {
          "PasswordPolicy": {
            "MinimumLength": 8,
            "RequireLowercase": true,
            "RequireNumbers": true,
            "RequireSymbols": false,
            "RequireUppercase": true,
          },
        },
        "Schema": [
          {
            "Mutable": false,
            "Name": "email",
            "Required": true,
          },
          {
            "Mutable": true,
            "Name": "preferred_username",
            "Required": true,
          },
          {
            "Mutable": false,
            "Name": "updated_at",
            "Required": true,
          },
          {
            "AttributeDataType": "Number",
            "Mutable": false,
            "Name": "created_at",
          },
        ],
        "SmsVerificationMessage": "The verification code to your new account is {####}",
        "UserPoolName": "test-user-pool-name",
        "UsernameAttributes": [
          "email",
        ],
        "VerificationMessageTemplate": {
          "DefaultEmailOption": "CONFIRM_WITH_CODE",
          "EmailMessage": "Thanks for signing up. Please verify your account by the verification code {####}.",
          "EmailSubject": "[PRD] You need to verify your email",
          "SmsMessage": "The verification code to your new account is {####}",
        },
      },
      "Type": "AWS::Cognito::UserPool",
      "UpdateReplacePolicy": "Delete",
    },
    "testuserpoolnametestuserpoolclientname68410DE1": {
      "Properties": {
        "AllowedOAuthFlows": [
          "implicit",
          "code",
        ],
        "AllowedOAuthFlowsUserPoolClient": true,
        "AllowedOAuthScopes": [
          "profile",
          "phone",
          "email",
          "openid",
          "aws.cognito.signin.user.admin",
        ],
        "CallbackURLs": [
          "https://example.com",
        ],
        "ClientName": "test-user-pool-client-name",
        "ExplicitAuthFlows": [
          "ALLOW_USER_PASSWORD_AUTH",
          "ALLOW_REFRESH_TOKEN_AUTH",
        ],
        "SupportedIdentityProviders": [
          "COGNITO",
        ],
        "UserPoolId": {
          "Ref": "testuserpoolname6617CE2F",
        },
      },
      "Type": "AWS::Cognito::UserPoolClient",
    },
    "testuserpoolnametestuserpooldomainnameBB32991B": {
      "Properties": {
        "Domain": "test-user-pool-domain-prefix",
        "UserPoolId": {
          "Ref": "testuserpoolname6617CE2F",
        },
      },
      "Type": "AWS::Cognito::UserPoolDomain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
